### ISSUE TITLE
AssertThrowsShouldUseGenericOverloadCheck may change type of expression

### ISSUE THREAD

**Issue by jamesqo** [xunit/xunit#1434]
If someone actually uses the return value of `Assert.Throws` and declares a variable with `var`, the inferred type of the variable may change from `Exception` to a subclass which is breaking.

```cs
// ex is Exception
var ex = Assert.Throws(typeof(ArgumentException), func);

// ex is ArgumentException
var ex = Assert.Throws<ArgumentException>(func);

// Anything which involves generic type inference from ex will also be affected
Console.WriteLine(new[] { ex }.GetType()); // 1) System.Exception[], 2) System.ArgumentException[]
```

/cc @ErikSchierboom

**Comment by ErikSchierboom** [comment]
Hmmm, how do you propose to handle this case?

**Comment by jamesqo** [comment]
@ErikSchierboom We could either upcast to Exception or leave out the diagnostic altogether. I'm leaving towards option 1.

**Comment by ErikSchierboom** [comment]
So we would rewrite to: `Exception ex = Assert.Throws<ArgumentException>(func);`?

**Comment by jamesqo** [comment]
@ErikSchierboom I was thinking `var ex = (Exception)Assert.Throws<ArgumentException>(func)`-- we want to make the changes as local/self-contained as possible so we don't keep having to handle edge cases.

Note that this issue isn't just limited to `var ex = ...`, it's basically anywhere `Assert.Throws` is being used as an expression and its type is being inferred. For a contrived example,

```cs
var arr = new[] { Assert.Throws<ArgumentException>(func) };
```

I just came up with a solid idea, though. We can simply always add the cast in the codefix, e.g. even for `Exception ex = Assert.Throws(typeof(...), func)` where it's not needed, we would still add the `(Exception)` cast. Then we can attach `Simplifier.Annotation` [like so](https://gitlab.com/jamesqo/MiscAnalyzers/blob/master/MiscAnalyzers/MiscAnalyzers/Internal/ExpressionSyntaxExtensions.cs#L68) to the cast expression, which causes Roslyn to remove the cast if it's redundant.

**Comment by ErikSchierboom** [comment]
@jamesqo That sounds like a solid plan! So the process is basically this:

1. `var ex = Assert.Throws(typeof(ArgumentException), func);`

2. `var ex = (Exception)Assert.Throws<ArgumentException>(func);`

3. `var ex = Assert.Throws<ArgumentException>(func);`

Where the last step is done by `Simplifier.Annotation`, correct?

**Comment by jamesqo** [comment]
@ErikSchierboom Yep. You may have to ensure `Assert.Throws` is being used as an expression and not a statement if you do that, though-- something like

```cs
(Exception)Assert.Throws<ArgumentException>(func);
```

by itself may be an error/may not be eligible for simplication. (I haven't tested that, though.)

**Comment by marcind** [comment]
I'm finding the example cited in https://github.com/xunit/xunit/issues/1434#issue-254538892 to be quite contrived. The value of the strongly typed `Throws<T>` is that you can write:

```csharp
var ex = Assert.Throws<AggregateException>(func);
var innerEx = Assert.Single(ex.InnerExceptions);
```

instead of

```csharp
var ex = Assert.Throws(typeof(AggregateException), func);
var innerEx = Assert.Single(((AggregateException)ex).InnerExceptions);
```

Unless I'm misreading, it sounds like the proposal is to offer `(Exception)Assert.Throws<ArgumentException>(func);` as a replacement, which to me feels like a step in the wrong direction. I'd like to see a real example where the current behavior has an adverse effect before changing anything.

**Comment by bradwilson** [comment]
Hello,

As part of the effort to clean up older issues, this is being closed due to age. If this is still applicable with v3, please open a new issue (with an attached v3 repro) so it can be re-evaluated and categorized.

Thanks!
